data-structures/ruby/doubly_linked_list/base.rb

```diff
@@ -0,0 +1,123 @@
+require_relative 'empty'
+require_relative 'node'
+require_relative 'doubly_linked_list_node'
+
+module DoublyLinkedList
+  class Base
+
+    include DoublyLinkedList
+    include Enumerable
+
+    attr_reader :head, :tail, :length
+    def initialize()
+      empty = DoublyLinkedList::Empty.new()
+      @head = empty
+      @tail = empty
+      @length = 0
+    end
+
+    def tail
+      @tail
+    end
+
+    def calculate_length
+      length = 0
+      placeholder = @head
+      until placeholder.is_a?(DoublyLinkedList::Empty) do
+        length += 1
+        placeholder = placeholder.next
+      end
+      length
+    end
+
+    def find_tail
+      placeholder = @head
+      if placeholder.is_a?(DoublyLinkedList::Empty)
+        @head
+      else
+        until placeholder.next.is_a?(DoublyLinkedList::Empty) do
+          placeholder.next
+        end
+        placeholder
+      end
+    end
+
+    def each(&block)
+      placeholder = @head
+      until placeholder.is_a?(DoublyLinkedList::Empty) do
+        block.call(placeholder.value)
+        placeholder = placeholder.next
+      end
+
+      self
+    end
+
+    def shift
+      head = @head
+      @head = @head.next
+      @length -= 1
+      head.value
+    end
+
+    def unshift(value)
+      node = DoublyLinkedListNode(value, @head)
+      if @head.is_a?(Empty)
```

Comment:
Type checking is a code smell.  Make `Node` and `EmptyNode` implement the same interface so that things that use them need not care which type the object is.